walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,96 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
 
 public final class FindMeetingQuery {
+  /**
+   * Given the currently booked events (time ranges and attendees) and
+   * a meeting request (duration and attendees), return a collection
+   * the of time ranges within which the meeting can be booked.
+   */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Collection<TimeRange> results = new ArrayList<TimeRange>();
+    Collection<String> attendees = request.getAttendees();
+    long duration = request.getDuration();
+
+    // Not possible to have an event longer than a full day
+    if (duration > TimeRange.WHOLE_DAY.duration()) {
+      return results;
+    }
+
+    // If no one's attending, there are no conflicts
+    if (attendees.isEmpty()) {
+      results.add(TimeRange.WHOLE_DAY);
+      return results;
+    }
+
+    ArrayList<TimeRange> invalidRanges = new ArrayList<TimeRange>();
+    for (Event event : events) {
+      Set<String> intersection = new HashSet<String>(attendees);
+
+      // If there are attendees in common between the request and this event
+      if(!Collections.disjoint(attendees, event.getAttendees())) {
+        // Requested meeting can't be at same time as this event
+        invalidRanges.add(event.getWhen());
+      }
+    }
+
+    if (invalidRanges.isEmpty()) {
+      results.add(TimeRange.WHOLE_DAY);
+      return results;
+    }
+
+    Collections.sort(invalidRanges, TimeRange.ORDER_BY_START);
+    
+    int invalidRangeStart;
+    int minInvalidRangeEnd;
+    int validRangeStart;
+    int validRangeEnd;
+
+    // If first event doesn't start at start of day, add time range beginning
+    // at start of day
+    if (invalidRanges.get(0).start() != TimeRange.START_OF_DAY) {
+      validRangeStart = TimeRange.START_OF_DAY;
+      validRangeEnd = invalidRanges.get(0).start();
+      TimeRange validRange = TimeRange.fromStartEnd(validRangeStart, validRangeEnd, false);
+      // Add to results if it's long enough
+      if (validRange.duration() >= duration) {
+        results.add(validRange);
+      }
+    }
+
+    invalidRangeStart = invalidRanges.get(0).start();
+    minInvalidRangeEnd = invalidRanges.get(0).end();
+    for (int i = 1; i < invalidRanges.size(); i++) {
```

Comment:
Could add another comment here describing your general strategy in this for loop